tests/test_client_session.py

```diff
@@ -467,6 +467,60 @@ async def create_connection(req, traces, timeout):
         c.__del__()
 
 
+@pytest.mark.parametrize("protocol", ["http", "https", "ws", "wss"])
+async def test_ws_connect_allowed_protocols(
+    create_session: Any,
+    create_mocked_conn: Any,
+    protocol: str,
+    ws_key: Any,
+    key_data: Any,
+) -> None:
+    resp = mock.create_autospec(aiohttp.ClientResponse)
+    resp.status = 101
+    resp.headers = {
+        hdrs.UPGRADE: "websocket",
+        hdrs.CONNECTION: "upgrade",
+        hdrs.SEC_WEBSOCKET_ACCEPT: ws_key,
+    }
+    resp.url = URL(f"{protocol}://example.com")
+    resp.cookies = SimpleCookie()
+    resp.start = mock.AsyncMock()
+
+    req = mock.create_autospec(aiohttp.ClientRequest, spec_set=True)
+    req_factory = mock.Mock(return_value=req)
+    req.send = mock.AsyncMock(return_value=resp)
+
+    session = await create_session(request_class=req_factory)
+
+    connections = []
+    original_connect = session._connector.connect
+
+    async def connect(req, traces, timeout):
+        conn = await original_connect(req, traces, timeout)
+        connections.append(conn)
+        return conn
+
+    async def create_connection(req, traces, timeout):
+        return create_mocked_conn()
+
+    connector = session._connector
+    with mock.patch.object(connector, "connect", connect), mock.patch.object(
+        connector, "_create_connection", create_connection
+    ), mock.patch.object(connector, "_release"), mock.patch(
+        "aiohttp.client.os"
+    ) as m_os:
+        m_os.urandom.return_value = key_data
+        await session.ws_connect(f"{protocol}://example.com")
+
+    # normally called during garbage collection.  triggers an exception
+    # if the connection wasn't already closed
+    for c in connections:
+        c.close()
+        del c
```

Comment:
FYI, `__del__()` is more or less predictable in CPython, but in PyPy it's not really guaranteed to be called at a specific moment in time. Calling `gc.collect()` might improve it, especially in tests, but might need several invocations.

Reply:
Do you think it better to switch it back to `__del__`? codeql complained about it

> Code scanning
> / CodeQL
> `__del__` is called explicitly
> The __del__ special method is called explicitly.
> [Show more details](https://github.com/aio-libs/aiohttp/security/code-scanning/599)

Reply:
Yeah, linters would normally complain and you'd have to ignore. However, I'd probably just stick a few calls to `gc.collect()`, maybe.

Reply:
https://github.com/aio-libs/aiohttp/pull/8513